src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProvider.java

```diff
@@ -0,0 +1,252 @@
+            LSPSemanticToken semanticToken = getSemanticToken(offset);
+            return semanticToken != null ? semanticToken.getElement() : super.findElementAt(offset);
+        } else {
+            return super.findElementAt(offset);
+        }
+    }
+
+    @Override
+    public PsiElement findElementAt(int offset, @NotNull Class<? extends Language> lang) {
+        return isSupported() ? findElementAt(offset) : super.findElementAt(offset, lang);
+    }
+
+    @Override
+    public PsiElement findElementAt(int offset, @NotNull Language language) {
+        return isSupported() ? findElementAt(offset) : super.findElementAt(offset, language);
+    }
+
+    @Override
+    public PsiReference findReferenceAt(int offset) {
+        if (isSupported()) {
+            LSPSemanticToken semanticToken = getSemanticToken(offset);
+            LSPSemanticTokenElementType elementType = semanticToken != null ? semanticToken.getElementType() : null;
+            return elementType == LSPSemanticTokenElementType.REFERENCE ? new LSPSemanticTokenPsiReference(semanticToken) : null;
+        } else {
+            return super.findReferenceAt(offset);
+        }
+    }
+
+    @Override
+    @Nullable
+    public PsiReference findReferenceAt(int offset, @NotNull Language language) {
+        return isSupported() ? findReferenceAt(offset) : super.findReferenceAt(offset, language);
+    }
+
+    // Element description
+
+    /**
+     * Returns the description for the specified element, optionally formatted as appropriate from the perspective of
+     * the provided reference element. For example, if both are elements are in the same file, there's no reason to
+     * include the target element's file name in the description, but if they're in different files, it's useful to
+     * know where the target element resides.
+     *
+     * @param element          the element for which a description should be returned
+     * @param referenceElement the reference element from which the description is being requested
+     * @return the element description
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public String getElementDescription(@NotNull PsiElement element, @Nullable PsiElement referenceElement) {
+        if (findElementAt(element.getTextOffset()) instanceof LSPSemanticTokenPsiElement semanticTokenElement) {
+            return semanticTokenElement.getDescription(referenceElement);
+        }
+        return null;
+    }
+
+    // Semantic tokens management
+
+    // Store the file's semantic tokens so that we have constant-time lookup of an element for a given offset
+    @Nullable
+    private Map<Integer, LSPSemanticToken> getSemanticTokensByOffset() {
+        PsiFile file = getSupportedFile();
+        if (file != null) {
+            // By caching the storage on the file this way, it's automatically evicted when the file changes
+            return CachedValuesManager.getCachedValue(file, new CachedValueProvider<>() {
```

Comment:
Hopefully you're already familiar with `CachedValuesManager`, but I'm using it here and one other place below to cache file-specific information that should expire when the `PsiFile` is rendered invalid by edits, file type changes, etc.